src/pipeline.cpp

```diff
@@ -0,0 +1,604 @@
+//? annotation2metadata pass
+//? force function attributes pass
+//? annotation remarks pass
+//? infer function attributes pass
+//? lower expect intrinsic pass
+//? warn missed transformations pass
+//* For vectorization
+//? loop unroll/jam after loop vectorization
+//? optimization remarks pass
+//? cse/cvp/instcombine/bdce/sccp/licm/unswitch after loop vectorization (
+// cleanup as much as possible before trying to slp vectorize)
+//? vectorcombine pass
+//* For optimization
+//? float2int pass
+//? lower constant intrinsics pass
+//? loop sink pass
+//? hot-cold splitting pass
+
+//Use for O1 and below
+void buildBasicPipeline(ModulePassManager &MPM, PassBuilder &PB, OptimizationLevel O, OptimizationOptions options) {
+// #ifdef JL_DEBUG_BUILD
+    addVerificationPasses(MPM);
+// #endif
+    invokePipelineStartCallbacks(MPM, PB, O);
+    MPM.addPass(ConstantMergePass());
+    if (!options.dump_native) {
+        MPM.addPass(CPUFeatures());
+        if (O.getSpeedupLevel() > 0) {
+            MPM.addPass(createModuleToFunctionPassAdaptor(InstSimplifyPass()));
+        }
+    }
+    {
+        FunctionPassManager FPM;
+        FPM.addPass(SimplifyCFGPass(basicSimplifyCFGOptions()));
+        if (O.getSpeedupLevel() > 0) {
+            FPM.addPass(SROAPass());
+            FPM.addPass(InstCombinePass());
+            FPM.addPass(EarlyCSEPass());
+        }
+        FPM.addPass(MemCpyOptPass());
+        MPM.addPass(createModuleToFunctionPassAdaptor(std::move(FPM)));
+    }
+    invokeEarlySimplificationCallbacks(MPM, PB, O);
+    MPM.addPass(AlwaysInlinerPass());
+    {
+        CGSCCPassManager CGPM;
+        invokeCGSCCCallbacks(CGPM, PB, O);
+        MPM.addPass(createModuleToPostOrderCGSCCPassAdaptor(std::move(CGPM)));
+    }
+    invokeOptimizerEarlyCallbacks(MPM, PB, O);
+    MPM.addPass(LowerSIMDLoop());
+    {
+        FunctionPassManager FPM;
+        {
+            LoopPassManager LPM;
+            invokeLateLoopOptimizationCallbacks(LPM, PB, O);
+            invokeLoopOptimizerEndCallbacks(LPM, PB, O);
+            FPM.addPass(createFunctionToLoopPassAdaptor(std::move(LPM)));
+        }
+        invokeScalarOptimizerCallbacks(FPM, PB, O);
+        invokeVectorizerCallbacks(FPM, PB, O);
+        MPM.addPass(createModuleToFunctionPassAdaptor(std::move(FPM)));
+    }
+    if (options.lower_intrinsics) {
+        //TODO no barrier pass?
+        {
+            FunctionPassManager FPM;
+            FPM.addPass(LowerExcHandlers());
+            FPM.addPass(GCInvariantVerifierPass(false));
+            MPM.addPass(createModuleToFunctionPassAdaptor(std::move(FPM)));
+        }
+        MPM.addPass(RemoveNI());
+        MPM.addPass(createModuleToFunctionPassAdaptor(LateLowerGC()));
+        MPM.addPass(FinalLowerGCPass());
+        MPM.addPass(LowerPTLSPass(options.dump_native));
+    } else {
+        MPM.addPass(RemoveNI());
+    }
+    MPM.addPass(LowerSIMDLoop()); // TODO why do we do this twice
```

Comment:
While translating our pass pipeline over to NewPM, I noticed that we're calling for LowerSIMDLoop to run twice in the pipeline if we're running at O1 or below. Is there a particular reason for this double run?

Reply:
No I think we should only run it once